core/rawdb/accessors_metadata_ext.go

```diff
@@ -0,0 +1,109 @@
+		return time.Time{}, err
+	}
+
+	return time.Unix(int64(lastRun), 0), nil
+}
+
+// WriteOfflinePruning writes a time marker of the last attempt to run offline pruning.
+// The marker is written when offline pruning completes and is deleted when the node
+// is started successfully with offline pruning disabled. This ensures users must
+// disable offline pruning and start their node successfully between runs of offline
+// pruning.
+func WriteOfflinePruning(db ethdb.KeyValueStore) error {
+	return writeCurrentTimeMarker(db, offlinePruningKey)
+}
+
+// ReadOfflinePruning reads the most recent timestamp of an attempt to run offline
+// pruning if present.
+func ReadOfflinePruning(db ethdb.KeyValueStore) (time.Time, error) {
+	return readTimeMarker(db, offlinePruningKey)
+}
+
+// DeleteOfflinePruning deletes any marker of the last attempt to run offline pruning.
+func DeleteOfflinePruning(db ethdb.KeyValueStore) error {
+	return db.Delete(offlinePruningKey)
+}
+
+// WritePopulateMissingTries writes a marker for the current attempt to populate
+// missing tries.
+func WritePopulateMissingTries(db ethdb.KeyValueStore) error {
+	return writeCurrentTimeMarker(db, populateMissingTriesKey)
+}
+
+// ReadPopulateMissingTries reads the most recent timestamp of an attempt to
+// re-populate missing trie nodes.
+func ReadPopulateMissingTries(db ethdb.KeyValueStore) (time.Time, error) {
+	return readTimeMarker(db, populateMissingTriesKey)
+}
+
+// DeletePopulateMissingTries deletes any marker of the last attempt to
+// re-populate missing trie nodes.
+func DeletePopulateMissingTries(db ethdb.KeyValueStore) error {
+	return db.Delete(populateMissingTriesKey)
+}
+
+// WritePruningDisabled writes a marker to track whether the node has ever run
+// with pruning disabled.
+func WritePruningDisabled(db ethdb.KeyValueStore) error {
+	return db.Put(pruningDisabledKey, nil)
+}
+
+// HasPruningDisabled returns true if there is a marker present indicating that
+// the node has run with pruning disabled at some pooint.
+func HasPruningDisabled(db ethdb.KeyValueStore) (bool, error) {
+	return db.Has(pruningDisabledKey)
+}
+
+// WriteAcceptorTip writes `hash` as the last accepted block that has been fully processed.
+func WriteAcceptorTip(db ethdb.KeyValueWriter, hash common.Hash) error {
+	return db.Put(acceptorTipKey, hash[:])
+}
+
+// ReadAcceptorTip reads the hash of the last accepted block that was fully processed.
+// If there is no value present (the index is being initialized for the first time), then the
+// empty hash is returned.
+func ReadAcceptorTip(db ethdb.KeyValueReader) (common.Hash, error) {
+	has, err := db.Has(acceptorTipKey)
+	if err != nil {
+		return common.Hash{}, err
+	} else if !has {
+		// If the index is not present on disk, the [acceptorTipKey] index has not been initialized yet.
+		return common.Hash{}, nil
+	}
+	h, err := db.Get(acceptorTipKey)
+	if err != nil {
+		return common.Hash{}, err
+	}
+	return common.BytesToHash(h), nil
```

Comment:
I think you should check the length of `h`. I've been bitten many times by false assumptions (of my own and of others) when using the `common` converters.

Reply:
Bundled in 1ccca3b0bdaf5fc3516a95d60596678a3ee78738.
Note these are simply code movements of existing code, but why not change them whilst we're at it indeed 👍